File: saksflyt/src/test/java/no/nav/melosys/saksflyt/steg/register/HentRegisteropplysningerTest.java
package no.nav.melosys.saksflyt.steg.register;

import java.time.LocalDate;

import io.getunleash.FakeUnleash;
import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.BehandlingTestFactory;
import no.nav.melosys.domain.Fagsak;
import no.nav.melosys.domain.FagsakTestFactory;
import no.nav.melosys.domain.kodeverk.Sakstyper;
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstema;
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper;
import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysninger;
import no.nav.melosys.domain.mottatteopplysninger.Soeknad;
import no.nav.melosys.domain.mottatteopplysninger.SøknadNorgeEllerUtenforEØS;
import no.nav.melosys.domain.mottatteopplysninger.data.Periode;
import no.nav.melosys.saksflytapi.domain.Prosessinstans;
import no.nav.melosys.service.behandling.BehandlingService;
import no.nav.melosys.service.persondata.PersondataFasade;
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerFactory;
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerRequest;
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerService;
import no.nav.melosys.service.saksbehandling.SaksbehandlingRegler;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class HentRegisteropplysningerTest {

    @Mock
    private RegisteropplysningerService registeropplysningerService;
    @Mock
    private BehandlingService behandlingService;
    @Mock
    private PersondataFasade persondataFasade;
    @Mock
    private SaksbehandlingRegler saksbehandlingRegler;

    private HentRegisteropplysninger hentRegisteropplysninger;

    @Captor
    private ArgumentCaptor<RegisteropplysningerRequest> requestCaptor;

    private final Behandling behandling = BehandlingTestFactory.builderWithDefaults().build();

    private FakeUnleash fakeUnleash = new FakeUnleash();

    @BeforeEach
    public void setUp() {
        RegisteropplysningerFactory registeropplysningerFactory = new RegisteropplysningerFactory(saksbehandlingRegler, fakeUnleash);
        hentRegisteropplysninger = new HentRegisteropplysninger(registeropplysningerService, behandlingService, saksbehandlingRegler, persondataFasade, registeropplysningerFactory);

        behandling.setId(222L);

        Fagsak fagsak = FagsakTestFactory.builder().medBruker().build();
        behandling.setFagsak(fagsak);
        behandling.setType(Behandlingstyper.FØRSTEGANG);

        when(behandlingService.hentBehandling(behandling.getId())).thenReturn(behandling);
    }

    @Test
    void utfør_hoppOverSteg() {
        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);
        Fagsak fagsak = FagsakTestFactory.builder().type(Sakstyper.FTRL).medBruker().build();
        behandling.setFagsak(fagsak);
        behandling.setTema(Behandlingstema.ARBEID_KUN_NORGE);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_hoppOverSteg_virksomhet() {
        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);

        Fagsak fagsak = FagsakTestFactory.builder().type(Sakstyper.FTRL).medVirksomhet().build();
        behandling.setFagsak(fagsak);
        behandling.setTema(Behandlingstema.ARBEID_KUN_NORGE);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_behandlingstemaUtsendtArbeidstaker_henterPeriodeFraSøknad() {
        String ident = "143545";
        when(persondataFasade.hentFolkeregisterident(FagsakTestFactory.BRUKER_AKTØR_ID)).thenReturn(ident);

        behandling.getFagsak().setType(Sakstyper.EU_EOS);
        behandling.setTema(Behandlingstema.UTSENDT_ARBEIDSTAKER);

        Periode periode = new Periode(LocalDate.now(), LocalDate.now().plusYears(2));
        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(new Soeknad());
        mottatteOpplysninger.getMottatteOpplysningerData().periode = periode;
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);


        hentRegisteropplysninger.utfør(prosessinstans);


        verify(registeropplysningerService).hentOgLagreOpplysninger(requestCaptor.capture());

        assertThat(requestCaptor.getValue())
            .extracting(RegisteropplysningerRequest::getBehandlingID, RegisteropplysningerRequest::getFnr, RegisteropplysningerRequest::getFom, RegisteropplysningerRequest::getTom)
            .containsExactly(behandling.getId(), ident, periode.getFom(), periode.getTom());
    }

    @Test
    void utfør_sakstypeFtrl_ingentingLagres() {
        behandling.setTema(Behandlingstema.YRKESAKTIV);
        behandling.getFagsak().setType(Sakstyper.FTRL);

        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(new SøknadNorgeEllerUtenforEØS());
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_sakstypeTrygdeavtale_ingentingLagres() {
        behandling.setTema(Behandlingstema.YRKESAKTIV);
        behandling.getFagsak().setType(Sakstyper.TRYGDEAVTALE);

        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(new SøknadNorgeEllerUtenforEØS());
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_sakstypeEøsOgUnntak_ingentingLagres() {
        behandling.setTema(Behandlingstema.A1_ANMODNING_OM_UNNTAK_PAPIR);
        behandling.getFagsak().setType(Sakstyper.EU_EOS);
        behandling.setType(Behandlingstyper.FØRSTEGANG);
        when(saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(behandling)).thenReturn(true);

        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(new SøknadNorgeEllerUtenforEØS());
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_sakstypeEøsOgIkkeYrkesaktiv_ingentingLagres() {
        behandling.setTema(Behandlingstema.IKKE_YRKESAKTIV);
        behandling.getFagsak().setType(Sakstyper.EU_EOS);
        behandling.setType(Behandlingstyper.FØRSTEGANG);
        when(saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(behandling)).thenReturn(true);

        MottatteOpplysninger mottatteOpplysninger = new MottatteOpplysninger();
        mottatteOpplysninger.setMottatteOpplysningerData(new SøknadNorgeEllerUtenforEØS());
        behandling.setMottatteOpplysninger(mottatteOpplysninger);

        Prosessinstans prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService, never()).hentOgLagreOpplysninger(any());
    }

    @Test
    void utfør_harIngenFlyt_henterIngenting() {
        behandling.setTema(Behandlingstema.TRYGDETID);
        behandling.getFagsak().setType(Sakstyper.EU_EOS);
        var prosessinstans = new Prosessinstans();
        prosessinstans.setBehandling(behandling);
        when(saksbehandlingRegler.harIngenFlyt(any(), any(), any(), any())).thenReturn(true);

        hentRegisteropplysninger.utfør(prosessinstans);

        verify(registeropplysningerService).hentOgLagreOpplysninger(requestCaptor.capture());
        assertThat(requestCaptor.getValue().getOpplysningstyper()).isEmpty();
    }
}

File: saksflyt/src/test/kotlin/no/nav/melosys/saksflyt/steg/register/HentRegisteropplysningerTest.kt
package no.nav.melosys.saksflyt.steg.register

import io.getunleash.FakeUnleash
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.shouldBe
import io.mockk.*
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.FagsakTestFactory
import no.nav.melosys.domain.fagsak
import no.nav.melosys.domain.forTest
import no.nav.melosys.domain.kodeverk.Sakstyper
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstema
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper
import no.nav.melosys.domain.mottatteopplysninger.SøknadNorgeEllerUtenforEØS
import no.nav.melosys.domain.mottatteopplysninger.mottatteOpplysningerForTest
import no.nav.melosys.domain.mottatteopplysninger.soeknad
import no.nav.melosys.domain.mottatteOpplysninger
import no.nav.melosys.saksflytapi.domain.ProsessStatus
import no.nav.melosys.saksflytapi.domain.ProsessType
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import no.nav.melosys.saksflytapi.domain.behandling
import no.nav.melosys.saksflytapi.domain.forTest
import no.nav.melosys.service.behandling.BehandlingService
import no.nav.melosys.service.persondata.PersondataFasade
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerFactory
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerRequest
import no.nav.melosys.service.registeropplysninger.RegisteropplysningerService
import no.nav.melosys.service.saksbehandling.SaksbehandlingRegler
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.time.LocalDate

@ExtendWith(MockKExtension::class)
class HentRegisteropplysningerTest {

    @MockK
    private lateinit var registeropplysningerService: RegisteropplysningerService

    @MockK
    private lateinit var behandlingService: BehandlingService

    @MockK
    private lateinit var persondataFasade: PersondataFasade

    @MockK
    private lateinit var saksbehandlingRegler: SaksbehandlingRegler

    private lateinit var hentRegisteropplysninger: HentRegisteropplysninger

    private val requestCaptor = slot<RegisteropplysningerRequest>()

    private val fakeUnleash = FakeUnleash()

    @BeforeEach
    fun setUp() {
        val registeropplysningerFactory = RegisteropplysningerFactory(saksbehandlingRegler, fakeUnleash)
        hentRegisteropplysninger = HentRegisteropplysninger(
            registeropplysningerService,
            behandlingService,
            saksbehandlingRegler,
            persondataFasade,
            registeropplysningerFactory
        )

        // Mock saksbehandlingRegler metoder
        every { saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(any()) } returns false
        every { saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(any(), any(), any()) } returns false
        every { saksbehandlingRegler.harIkkeYrkesaktivFlyt(any()) } returns false
        every { saksbehandlingRegler.harIngenFlyt(any(), any(), any(), any()) } returns false

        // Mock persondataFasade
        every { persondataFasade.hentFolkeregisterident(any()) } returns "12345678901"

        // Mock registeropplysningerService
        every { registeropplysningerService.hentOgLagreOpplysninger(any()) } just Runs
    }

    private fun lagBehandling(
        init: no.nav.melosys.domain.BehandlingTestFactory.BehandlingTestBuilder.() -> Unit = {}
    ): Behandling = Behandling.forTest {
        id = 222L
        type = Behandlingstyper.FØRSTEGANG
        fagsak { medBruker() }
        init()
    }.also { behandling ->
        every { behandlingService.hentBehandling(behandling.id) } returns behandling
    }

    @Test
    fun `utfør skal hoppe over steg`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.ARBEID_KUN_NORGE
            fagsak {
                type = Sakstyper.FTRL
                medBruker()
            }
        }

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal hoppe over steg for virksomhet`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.ARBEID_KUN_NORGE
            fagsak {
                type = Sakstyper.FTRL
                medVirksomhet()
            }
        }

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal hente periode fra søknad når behandlingstema er utsendt arbeidstaker`() {
        val ident = "143545"
        every { persondataFasade.hentFolkeregisterident(FagsakTestFactory.BRUKER_AKTØR_ID) } returns ident

        val periodeFom = LocalDate.now()
        val periodeTom = LocalDate.now().plusYears(2)
        val behandling = lagBehandling {
            tema = Behandlingstema.UTSENDT_ARBEIDSTAKER
            fagsak {
                type = Sakstyper.EU_EOS
                medBruker()
            }
            mottatteOpplysninger {
                soeknad {
                    periode(periodeFom, periodeTom)
                }
            }
        }

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify { registeropplysningerService.hentOgLagreOpplysninger(capture(requestCaptor)) }

        requestCaptor.captured.run {
            behandlingID shouldBe behandling.id
            fnr shouldBe ident
            fom shouldBe periodeFom
            tom shouldBe periodeTom
        }
    }

    @Test
    fun `utfør skal ikke lagre noe når sakstype er FTRL`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.YRKESAKTIV
            fagsak {
                type = Sakstyper.FTRL
                medBruker()
            }
            mottatteOpplysninger = mottatteOpplysningerForTest {
                mottatteOpplysningerData = SøknadNorgeEllerUtenforEØS()
            }
        }

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal ikke lagre noe når sakstype er TRYGDEAVTALE`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.YRKESAKTIV
            fagsak {
                type = Sakstyper.TRYGDEAVTALE
                medBruker()
            }
            mottatteOpplysninger = mottatteOpplysningerForTest {
                mottatteOpplysningerData = SøknadNorgeEllerUtenforEØS()
            }
        }

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal ikke lagre noe når sakstype er EØS og unntak`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.A1_ANMODNING_OM_UNNTAK_PAPIR
            type = Behandlingstyper.FØRSTEGANG
            fagsak {
                type = Sakstyper.EU_EOS
                medBruker()
            }
            mottatteOpplysninger = mottatteOpplysningerForTest {
                mottatteOpplysningerData = SøknadNorgeEllerUtenforEØS()
            }
        }
        every { saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(behandling) } returns true

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal ikke lagre noe når sakstype er EØS og ikke yrkesaktiv`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.IKKE_YRKESAKTIV
            type = Behandlingstyper.FØRSTEGANG
            fagsak {
                type = Sakstyper.EU_EOS
                medBruker()
            }
            mottatteOpplysninger = mottatteOpplysningerForTest {
                mottatteOpplysningerData = SøknadNorgeEllerUtenforEØS()
            }
        }
        every { saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(behandling) } returns true

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal ikke lagre noe når det er Årsavregning for EØS-Pensjonist`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.PENSJONIST
            type = Behandlingstyper.ÅRSAVREGNING
            fagsak {
                type = Sakstyper.EU_EOS
                medBruker()
            }
            mottatteOpplysninger = mottatteOpplysningerForTest {
                mottatteOpplysningerData = SøknadNorgeEllerUtenforEØS()
            }
        }
        every { saksbehandlingRegler.harRegistreringUnntakFraMedlemskapFlyt(behandling) } returns true

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify(exactly = 0) { registeropplysningerService.hentOgLagreOpplysninger(any()) }
    }

    @Test
    fun `utfør skal hente ingenting når har ingen flyt`() {
        val behandling = lagBehandling {
            tema = Behandlingstema.TRYGDETID
            fagsak {
                type = Sakstyper.EU_EOS
                medBruker()
            }
        }
        every { saksbehandlingRegler.harIngenFlyt(any(), any(), any(), any()) } returns true

        val prosessinstans = Prosessinstans.forTest {
            type = ProsessType.OPPRETT_SAK
            status = ProsessStatus.KLAR
            behandling { id = behandling.id }
            this.behandling = behandling
        }


        hentRegisteropplysninger.utfør(prosessinstans)


        verify { registeropplysningerService.hentOgLagreOpplysninger(capture(requestCaptor)) }
        requestCaptor.captured.opplysningstyper.shouldBeEmpty()
    }
}
